### PR TITLE
Handle services of type ExternalName by returning a CNAME

### PR DESCRIPTION
**Pull request by smarterclayton**
Compliance with Kube 1.4

@ncdc [test]

### CODE REVIEW

**Comment by ncdc** [comment]
@knobunc @liggitt

**Comment by ncdc** [comment]
Wow that took me way too long to figure out the bash craziness in the extended dns test for verifying that dig is working correctly :p.

LGTM

File: pkg/dns/serviceresolver.go [hunk 1]
```diff
@@ -140,8 +140,13 @@ func (b *ServiceResolver) Records(dnsName string, exact bool) ([]msg.Service, er
 
 		// if has a portal IP and looking at svc
 		if svc.Spec.ClusterIP != kapi.ClusterIPNone && !retrieveEndpoints {
+			hostValue := svc.Spec.ClusterIP
+			if svc.Spec.Type == kapi.ServiceTypeExternalName {
```
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: is this only valid if `len(svc.Spec.ClusterIP) == 0`?

**Comment by smarterclayton** [comment]
I blame Tim Hockin.  He's crazy, and I didn't want to rewrite those.

On Nov 4, 2016, at 3:49 PM, Andy Goldstein <notifications@github.com> wrote:

Wow that took me way too long to figure out the bash craziness in the
extended dns test for verifying that dig is working correctly :p.

LGTM

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11790#issuecomment-258530684>,
or mute
the thread

File: pkg/dns/serviceresolver.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func (b *ServiceResolver) Records(dnsName string, exact bool) ([]msg.Service, er
 		}
 
 		// no clusterIP and not headless, no DNS
-		if len(svc.Spec.ClusterIP) == 0 {
+		if len(svc.Spec.ClusterIP) == 0 && svc.Spec.Type != kapi.ServiceTypeExternalName {
```
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: is there any security implication to letting the user control the CNAME that comes back from a DNS query?

Comment (smarterclayton) [inline pkg/dns/serviceresolver.go]: You can already set up a proxy at that name that points to any server.  TLS only trusts the name the user gives, not the other way around.  Also, you trust the name and the thing the name points to already.

File: pkg/dns/serviceresolver.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/dns/serviceresolver.go]: Yes, but type dominates clusterIP

File: pkg/dns/serviceresolver.go (hunk 2 above)
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: ok

File: pkg/dns/serviceresolver.go (hunk 1 above)
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: ok

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[merge] (test passed)

**Pull request merged**